sdk/pinocchio/src/sysvars/slot_hashes.rs

```diff
@@ -0,0 +1,1013 @@
+/// to caller whether to use MAX_ENTRIES or to use a call such as `get_entry_count_from_slice_unchecked`
+#[inline(always)]
+pub unsafe fn get_hash_from_slice_unchecked(
+    data: &[u8],
+    target_slot: Slot,
+    num_entries: usize,
+) -> Option<&[u8; HASH_BYTES]> {
+    position_from_slice_binary_search_unchecked(data, target_slot, num_entries).map(|index| {
+        let entry_offset = NUM_ENTRIES_SIZE + index * ENTRY_SIZE;
+        let hash_offset = entry_offset + SLOT_SIZE;
+        let hash_bytes = data.get_unchecked(hash_offset..(hash_offset + HASH_BYTES));
+        &*(hash_bytes.as_ptr() as *const [u8; HASH_BYTES])
+    })
+}
+
+/// Gets a reference to the `SlotHashEntry` at a specific index from a raw byte slice **without validation**.
+///
+/// # Safety
+/// Caller must guarantee `data` contains a valid `SlotHashes` structure and that `index` is less than the entry count derived from the data's prefix.
+#[inline(always)]
+pub unsafe fn get_entry_from_slice_unchecked(data: &[u8], index: usize) -> &SlotHashEntry {
+    let entry_offset = NUM_ENTRIES_SIZE + index * ENTRY_SIZE;
+    let entry_bytes = data.get_unchecked(entry_offset..(entry_offset + ENTRY_SIZE));
+    &*(entry_bytes.as_ptr() as *const SlotHashEntry)
+}
+
+/// Iterator over the entries in `SlotHashes`.
+///
+/// Yields references `&'s SlotHashEntry` tied to the lifetime `'s` of the borrow
+/// of the `SlotHashes` instance.
+pub struct SlotHashesIterator<'s, T>
+where
+    T: Deref<Target = [u8]>,
+{
+    slot_hashes: &'s SlotHashes<T>,
+    current_index: usize,
+}
+
+// Implement Iterator trait for the custom iterator struct
+// TODO: trait extension for unchecked Iterator::next()
+impl<'s, T> Iterator for SlotHashesIterator<'s, T>
+where
+    T: Deref<Target = [u8]>,
+{
+    type Item = &'s SlotHashEntry;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        // Use safe get_entry method from SlotHashes
+        let entry = self.slot_hashes.get_entry(self.current_index);
+        if entry.is_some() {
+            self.current_index += 1;
+        }
+        entry
+    }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        let remaining = self.slot_hashes.len().saturating_sub(self.current_index);
+        (remaining, Some(remaining))
+    }
+}
+
+// Implement ExactSizeIterator as we know the exact length
+impl<T> ExactSizeIterator for SlotHashesIterator<'_, T> where T: Deref<Target = [u8]> {}
+
+// Implement IntoIterator for references to SlotHashes
+// This allows using `for entry in &slot_hashes { ... }`
+impl<'s, T> IntoIterator for &'s SlotHashes<T>
+where
+    T: Deref<Target = [u8]>,
+{
+    type Item = &'s SlotHashEntry;
+    type IntoIter = SlotHashesIterator<'s, T>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        SlotHashesIterator {
+            slot_hashes: self,
+            current_index: 0,
+        }
+    }
+}
+
+#[inline(always)]
+unsafe fn core_binary_search<FFirstSlot, FSlotAtIndex>(
```

Comment:
It seems wrong to mark this function as `unsafe` when the safety check applies to a specific instance of the `get_slot_at_index` implementation passed into this function.

I think this function does not need to be `unsafe`.
It is when the `get_slot_at_index` uses unsafe code, this is where the safety check applies.